src/coreclr/src/jit/lower.cpp

```diff
@@ -2986,6 +3017,175 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
+                        lclNum, fieldLclNum);
+                lclVar->ChangeType(fieldDsc->lvType);
+                canEnregister = true;
+            }
+        }
+        if (!canEnregister)
+        {
+            // TODO-1stClassStructs: We can no longer promote or enregister this struct,
+            // since it is referenced as a whole.
+            comp->lvaSetVarDoNotEnregister(lclNum DEBUGARG(Compiler::DNER_VMNeedsStackAddr));
+        }
+    }
+
+    var_types regType = varDsc->GetRegisterType(lclVar);
+    assert((regType != TYP_STRUCT) && (regType != TYP_UNKNOWN));
+    // Note: regType could be a small type, in this case return will generate an extension move.
+    lclVar->ChangeType(regType);
+    if (varTypeUsesFloatReg(ret) != varTypeUsesFloatReg(lclVar))
+    {
+        GenTreeUnOp* bitcast = new (comp, GT_BITCAST) GenTreeOp(GT_BITCAST, ret->TypeGet(), lclVar, nullptr);
+        ret->gtOp1           = bitcast;
+        BlockRange().InsertBefore(ret, bitcast);
+        ContainCheckBitCast(bitcast);
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// LowerCallStruct: Lowers a call node that returns a stuct.
+//
+// Arguments:
+//     call - The call node to lower.
+//
+// Note: it transforms the call's user.
+//
+void Lowering::LowerCallStruct(GenTreeCall* call)
+{
+    assert(!comp->compDoOldStructRetyping());
+    CORINFO_CLASS_HANDLE        retClsHnd = call->gtRetClsHnd;
+    Compiler::structPassingKind howToReturnStruct;
+    var_types                   returnType = comp->getReturnTypeForStruct(retClsHnd, &howToReturnStruct);
+    assert(!varTypeIsStruct(returnType) && returnType != TYP_UNKNOWN);
+    call->gtType = genActualType(returnType); // the callee normalizes small return types.
+
+    LIR::Use callUse;
+    if (BlockRange().TryGetUse(call, &callUse))
+    {
+        GenTree* user = callUse.User();
+        switch (user->OperGet())
+        {
+            case GT_RETURN:
+            case GT_STORE_LCL_VAR:
+                // Leave as is, the user will handle it.
+                break;
+
+            case GT_STORE_BLK:
+            case GT_STORE_OBJ:
+            {
+                GenTreeBlk* storeBlk = user->AsBlk();
+#ifdef DEBUG
+                unsigned storeSize = storeBlk->GetLayout()->GetSize();
+                assert(storeSize == genTypeSize(returnType));
+#endif // DEBUG
+                // For `STORE_BLK<2>(dst, call struct<2>)` we will have call retyped as `int`,
+                // but `STORE` has to use the unwidened type.
+                user->ChangeType(returnType);
+                user->SetOper(GT_STOREIND);
+            }
+            break;
+
+            case GT_STOREIND:
+                assert(user->TypeIs(TYP_SIMD8, TYP_REF));
+                user->ChangeType(returnType);
+                user->SetOper(GT_STOREIND);
```

Comment:
Why do we need this? It's already `GT_STOREIND` right?

Reply:
Thanks for catching that, fixed.